lib/galaxy/tool_util/linters/inputs.py

```diff
@@ -29,22 +29,31 @@ def lint_inputs(tool_xml, lint_ctx):
         if not is_valid_cheetah_placeholder(param_name):
             lint_ctx.warn("Param input [%s] is not a valid Cheetah placeholder.", param_name)
 
+        if param_type not in ["text", "integer", "float", "color", "boolean",
+                              "genomebuild", "library_data", "select",
+                              "data_column", "hidden", "hidden_data",
+                              "baseurl", "file", "data", "drill_down",
+                              "group_tag", "data_collection"]:
+            lint_ctx.error("Param input [%s] has an invalid type", param_name)
+
         if param_type == "data":
             if "format" not in param_attrib:
                 lint_ctx.warn("Param input [%s] with no format specified - 'data' format will be assumed.", param_name)
-
-        if param_type == "select":
+        elif param_type == "select":
             dynamic_options = param.get("dynamic_options", None)
             if dynamic_options is None:
                 dynamic_options = param.find("options")
 
             select_options = _find_with_attribute(param, 'option', 'value')
             if any(['value' not in option.attrib for option in select_options]):
-                lint_ctx.error("Option without value")
+                lint_ctx.error("Select [%s] has option without value", param_name)
 
             if dynamic_options is None and len(select_options) == 0:
                 message = "No options defined for select [%s]" % param_name
                 lint_ctx.warn(message)
+        elif param_type in ["hidden", "hidden_data"]:
```

Comment:
I don't remember what those were good for, are you 100% certain they always require a value ?
Both parameters seem to be initialized with `None` as their value.

Reply:
Not at all for `hidden_data`. 

For `hidden` the tool can't be executed (the execute button just does nothing). But you are right, this might be caused by a bug elsewhere.

Reply:
If you figure this out we should add some minimal documentation. Just searching the codebase I'm still not certain how they work exactly.

Reply:
Counter-examples:
- for `hidden`:
  - https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/tools/imp_exp/exp_history_to_archive.xml (this is a special tool though)
  - https://github.com/bgruening/galaxytools/blob/master/chemicaltoolbox/pgchem/pgchem/db_similarity_search.xml#L41
- for `hidden_data`: https://github.com/galaxyproject/tools-iuc/blob/master/tool_collections/cufflinks/cufflinks/cufflinks_wrapper.xml#L135

Reply:
Thanks for looking this up @nsoranzo . Lets see if those tools can be executed. I re-tested my example from above. If one adds `optional="true"` the tool can be executed.  

The same holds for the cufflinks example (which has `hidden_data`), i.e. removing the `optional="true"` leads to a tool that can not be executed. 

For the pgchem there are two hidden inputs in two when branches and therefor can not be executed at all (it also has a typo which make it invalid ` </outputs>>`). 

I also had the impression that the tool form jumped a bit when I hit the execute button ... maybe it tries to scroll to the hidden input and show an error message? 

So, the question is if we want to extend the linting in the suggested way, or if we decide to call it a bug that such tools can not be executed. 

Since running the test tool via planemo works in any case (independent of a value or optional input) and also the history export tool works I tend to the latter and drop (most of) this PR... but I have no idea where to look for the bug. I guess its somewhere in the JS.

Reply:
> but I have no idea where to look for the bug. I guess its somewhere in the JS.

If some older version of Galaxy can run them, you can try to `git bisect` to find when this broke.

Reply:
Tested 16.04, 17.01, 18.01, and 19.01 (using planemo serve): the test tool can not be executed in any of them. 

So I add a check for optional and then it should be fine.

Reply:
So the client prevents submission here (and that also explains why `optional="true"` works): https://github.com/galaxyproject/galaxy/blob/01ed5b92a54e22578893256b7ded0a86e236a5b3/client/src/mvc/tool/tool-form.js#L340

Since you can submit the tool via the API (that's why planemo test works) I would suggest removing that check for hidden parameter as a quick fix. Even better would be not sending out hidden parameters at all, I don't think the client needs to see these at all ... and in fact that may leak "secret"-y values. Not that that would be the right mechanism for secrets anyway.

Reply:
OK. Seems that my guess on the JS change was wrong or not stylish enough :) suggestions? 

Can edit the commits later on if desired... removing the reverted ones...